File: app/src/main/java/com/example/nasmovie/data/db/FavoriteDao.java
package com.example.nasmovie.data.db;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.example.nasmovie.data.model.Favorite;

import java.util.List;

/**
 * 收藏DAO
 */
@Dao
public interface FavoriteDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insert(Favorite favorite);

    @Delete
    void delete(Favorite favorite);

    @Query("DELETE FROM favorite WHERE movieId = :movieId")
    void deleteByMovieId(String movieId);

    @Query("DELETE FROM favorite")
    void deleteAll();

    @Query("SELECT * FROM favorite WHERE movieId = :movieId")
    Favorite getByMovieId(String movieId);

    @Query("SELECT * FROM favorite ORDER BY addTime DESC")
    List<Favorite> getAll();

    @Query("SELECT movieId FROM favorite")
    List<String> getAllMovieIds();

    @Query("SELECT EXISTS(SELECT 1 FROM favorite WHERE movieId = :movieId)")
    boolean isFavorite(String movieId);

    @Query("SELECT COUNT(*) FROM favorite")
    int getCount();
}

File: app/src/main/java/com/example/nasmovie/data/db/FavoriteDao.kt
package com.example.nasmovie.data.db

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.example.nasmovie.data.model.Favorite

/**
 * 收藏DAO
 */
@Dao
interface FavoriteDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(favorite: Favorite): Long

    @Delete
    suspend fun delete(favorite: Favorite)

    @Query("DELETE FROM favorite WHERE movieId = :movieId")
    suspend fun deleteByMovieId(movieId: String)

    @Query("DELETE FROM favorite")
    suspend fun deleteAll()

    @Query("SELECT * FROM favorite WHERE movieId = :movieId")
    suspend fun getByMovieId(movieId: String): Favorite?

    @Query("SELECT * FROM favorite ORDER BY addTime DESC")
    suspend fun getAll(): List<Favorite>

    @Query("SELECT movieId FROM favorite")
    suspend fun getAllMovieIds(): List<String>

    @Query("SELECT EXISTS(SELECT 1 FROM favorite WHERE movieId = :movieId)")
    suspend fun isFavorite(movieId: String): Boolean

    @Query("SELECT COUNT(*) FROM favorite")
    suspend fun getCount(): Int

    @Query("DELETE FROM favorite WHERE movieId IN (:movieIds)")
    suspend fun deleteByMovieIds(movieIds: List<String>)

    @Query("DELETE FROM favorite WHERE movieId IN (SELECT id FROM movie WHERE serverId = :serverId)")
    suspend fun deleteByServerId(serverId: Long)
}